Frameworks/Foundation/NSInvocationOperation.mm

```diff
@@ -14,30 +14,131 @@
 //
 //******************************************************************************
 
-#import <StubReturn.h>
-#import <Starboard.h>
 #import <Foundation/NSInvocationOperation.h>
+#import <Starboard/SmartTypes.h>
+#import <NSLogging.h>
+#import <memory>
 
 NSString* const NSInvocationOperationVoidResultException = @"NSInvocationOperationVoidResultException";
 NSString* const NSInvocationOperationCancelledException = @"NSInvocationOperationCancelledException";
 
+@interface NSInvocationOperation () {
+    StrongId<NSException> _caughtException;
+    StrongId<id> _result;
+}
+@end
+
 @implementation NSInvocationOperation
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (instancetype)initWithTarget:(id)target selector:(SEL)sel object:(id)arg {
-    UNIMPLEMENTED();
-    return StubReturn();
+    NSMethodSignature* methodSignature = [target methodSignatureForSelector:sel];
+    if (!methodSignature) {
+        [self release];
+        return nil;
+    }
+
+    if (methodSignature.numberOfArguments > 3) {
+        [self release];
+        NSTraceError(L"NSInvocationOperation",
+                     @"%hs: method '%hs' requires %u arguments (max 1)",
+                     __PRETTY_FUNCTION__,
+                     sel_getName(sel),
+                     methodSignature.numberOfArguments - 2);
+        return nil;
+    }
+
+    NSInvocation* invocation = [NSInvocation invocationWithMethodSignature:methodSignature];
+    invocation.target = target;
+    invocation.selector = sel;
+    if (methodSignature.numberOfArguments > 2) {
+        [invocation setArgument:&arg atIndex:2];
+    }
+
+    return [self initWithInvocation:invocation];
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
-- (instancetype)initWithInvocation:(NSInvocation*)inv {
-    UNIMPLEMENTED();
-    return StubReturn();
+- (instancetype)initWithInvocation:(NSInvocation*)invocation {
+    if (!invocation) {
+        [self release];
+        return nil;
+    }
+
+    if (self = [super init]) {
+        _invocation = [invocation retain];
+        [_invocation retainArguments];
+    }
+    return self;
+}
+
+- (void)dealloc {
+    [_invocation release];
+    [super dealloc];
 }
 
+/**
+ @Status Interoperable
+*/
+- (id)result {
+    @synchronized(self) {
+        if (!self.finished) {
+            return nil;
+        }
+
+        if (self.cancelled) {
+            [NSException raise:NSInvocationOperationCancelledException format:@"%hs: %@ was cancelled", __PRETTY_FUNCTION__, self];
+        }
+
+        if (_caughtException) {
+            [_caughtException raise];
+        }
+
+        if (_invocation.methodSignature.methodReturnLength == 0) {
+            [NSException raise:NSInvocationOperationVoidResultException format:@"%hs: %@ has a void result", __PRETTY_FUNCTION__, self];
+        }
+
+        return [[_result retain] autorelease];
```

Comment:
`_result` is a member. If we didn't retain+autorelease, this would be potentially fatal:

```objc
id r = [operation result];
[operation release]; // remove last reference
[r anything]; // boom: last owning reference was in operation
```